File: poh3/poh3.swift
//usr/bin/env swift $0 $@;exit

//let M=getInt(),N=getInt()
let M=Int(readLine()!)!,N=Int(readLine()!)!
var num=[Int](repeating:0,count:N)
var cost=[Int](repeating:0,count:N)
var total=0
var i=0,j=0

for j in 0..<N {
	let a=readLine()!.characters.split{$0==" "}
	//let q=getInt()
	//let r=getInt()
	let q=Int(String(a[0]))!
	let r=Int(String(a[1]))!
	total+=q
	num[j]=q
	cost[j]=r
}
var bag=[Int](repeating:0,count:total+1)
for i in 1...total {bag[i] = -1}
var t=0
for j in 0..<N {
	t+=num[j]
	for i in stride(from:t,through:num[j],by:-1) {
		if bag[i-num[j]]>=0 {
			let x=bag[i-num[j]]+cost[j]
			if bag[i]<0||bag[i]>x {
				bag[i]=x
			}
		}
	}
}
j = -1
for i in M...total { if j<0||(bag[i]>=0&&j>bag[i]) {j=bag[i]}}
print(j)

File: poh3/poh3.c
//usr/bin/env picoc $0 - $@;exit
#include <stdio.h>
int num[50],cost[50];
int bag[500000];

int main(){
	int M,N,q,r,total=0,i,j;
	scanf("%d%d",&M,&N);
	for(total=j=0;j<N;j++){
		scanf("%d%d",&q,&r);
		total+=q;
		num[j]=q;cost[j]=r;
	}
	//std::vector<int>bag(total+1);
	for(i=1;i<=total;i++)bag[i]=-1;
	for(j=0;j<N;j++){
		for(i=total;i>=num[j];i--){
			if(bag[i-num[j]]>=0){
				if(bag[i]<0||bag[i]>bag[i-num[j]]+cost[j]){
					bag[i]=bag[i-num[j]]+cost[j];
				}
			}
		}
	}
	j=-1;
	for(i=M;i<=total;i++)if(j<0||(bag[i]>=0&&j>bag[i]))j=bag[i];
	printf("%d\n",j);
	return 0;
}
